File: src/types/reaction.ts
/**
 * representation of a note reaction on the server side
 */
export type Reaction = {
  id: string;
  note: string;
  user: string;
  reactionType: ReactionType;
};

export type ReactionState = Reaction[];

export type ReactionType = "thinking" | "heart" | "like" | "dislike" | "joy" | "celebration" | "poop" | "tada" | "applause";

export const REACTION_EMOJI_MAP = new Map<ReactionType, string>([
  ["thinking", "🤔"],
  ["heart", "💖"],
  ["like", "👍"],
  ["dislike", "👎"],
  ["joy", "😂"],
  ["celebration", "🥳"],
  ["poop", "💩"],
]);

File: server/src/database/types/reaction.go
package types

import (
	"encoding/json"
	"errors"
)

type Reaction string

const (
	ReactionThinking    Reaction = "thinking"
	ReactionHeart       Reaction = "heart"
	ReactionLike        Reaction = "like"
	ReactionDislike     Reaction = "dislike"
	ReactionJoy         Reaction = "joy"
	ReactionCelebration Reaction = "celebration"
	ReactionPoop        Reaction = "poop"
)

func (reaction *Reaction) UnmarshalJSON(b []byte) error {
	var s string
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	unmarshalledReaction := Reaction(s)
	switch unmarshalledReaction {
	case ReactionThinking, ReactionHeart, ReactionLike, ReactionDislike, ReactionJoy, ReactionCelebration, ReactionPoop:
		*reaction = unmarshalledReaction
		return nil
	}
	return errors.New("invalid reaction")
}
